### DISCUSSION TITLE
Use Case for "Controllers"?

### DISCUSSION THREAD

**Discussion by 64Spaces** [kuestess/homebridge-platform-insteonlocal#249] [General]
By the way Scott, I sent you another donation today. Thank you for offering and working on this Homebridge Platform.

After letting my system run InsteonLocal on autopilot for a couple of years, I decided to take advantage of the recent Smarthome shutdown and upgrade my system. I actually had several "pre" Dual Band devices around the house and some were sort of flaky. So I purchased some eBay devices, changed to a PLM from a Hub, and updated InsteonLocal. The current enhancments and tweaking of InsteonLocal are great. I'm using the GroupMember feature much more and the new "Momentary' feature is working well.  I''m still studying the Target KeyPad Button feature.

So my question is around the Controllers setting. I understand the GroupID feature, but does anyone have a typical use case for the Controllers setting? Thanks so much, Ruben

*edited to change GroupID to GroupMember

**Comment by kuestess** [comment]
@64Spaces Many thanks for the donation, much appreciated.  The 'controllers' feature was contributed by @rasod - see here for his specific use case: [https://github.com/kuestess/homebridge-platform-insteonlocal/pull/35](https://github.com/kuestess/homebridge-platform-insteonlocal/pull/35).  Essentially you can have devices check status when a message from a 'controller' is defined - theoretically, those intsteon messages get caught by the event listener and status updated accordingly, however in certain cases those messages get dropped when there is a lot of Insteon traffic.

**Comment by rasod** [comment]
@64Spaces I added the controller feature cause I have many plug-in dimmers that are controlled by switchlincs. When the switchlinc is turned on/off/dim the plug-in dimmers turn on/off/dim.

When the switchlinc is turned on an event is sent to the hub and the plug-in updates for that one device. However the events from the plug-in dimmers turning on may not be seen by the hub so the plug-in doesn’t know the lights turned on, so HK gets out of sync.

The controller feature basically tells the plug that when it sees an event for device A (the controller) go check the status of devices B,C,D (the responders) just to keep everything in sync. Little overkill but just a suspenders and belt approach.

**Comment by 64Spaces** [comment]
Perfect, thank you @rasod and @kuestess. I have been experimenting with this tonight, but I have not found a solution. I added the "E" to the "controllers" device as an experiment. Is there a way to designate a button on a keypad to be the controller?

I would rather use "controllers". For example, I have Button E on a Keypad that controls a Fan Wall Switch. Both are cross-linked. When using "controllers", pressing the button does not sync HomeKit to show the fan is on. But if I create a device for the button, and use groupMembers combined with groupID, the syncing will work. But I do not really need or want to see a device for Button Fan in HomeKit. Thanks so much

This does not sync:

                "name": "Fan",
                "deviceID": "AABBCC",
                "controllers": "XXYYZZ:E",
                "dimmable": "no",
                "deviceType": "switch"

But this syncs:

                "name": "Button Fan",
                "deviceID": "XXYYZZ",
                "deviceType": "scene",
                "groupMembers": "AABBCC",
                "keypadbtn": "E",
                "groupID": "19"

**Comment by 64Spaces** [comment]
Just to circle back after various tests configurations, I have not been able to. get "controllers" to work with keypad buttons. I tried various line items for the devices and was not successful with any variation. For now, I'll use "groupMembers" in a Keybad Button definition. Please let me know if I missed something.

**Reply by THX723** [reply]
"controllers" is an array, so you'll need to encapsulate it with [brackets], even if it's just 1 device defined.
For example:

"controllers": ["XXYYZZ"],

or for 2 ore more device...

"controllers": ["XXYYZZ", "AABBCC"],

**Reply by 64Spaces** [reply]
Thank you, but still no joy. I think the issue is that a button on a keypad cannot be designated as a "controller".

**Reply by THX723** [reply]
I suspect you may have a different interpretation of what "controllers" is used for. I don't blame you. Even having been involved with this for years, I forget after not touching this in a year. It took me a few re-reads to remember what and how each parameters are suppose to work.

If I remember correctly, "controllers" are uses to update HomeKit device states, so they are in sync with the actual state of the device(s) that were indirectly controlled by another switch. For example, using a Keypad to turn on a Lamp module would obviously turn on the lamp, but in HomeKit, that lamp would not have been shown as ON. Defining Keypad "controller" with the lamp module would solve that problem.

**Reply by ctgreybeard** [reply]
What I have observed is that when you designated a device as a Controller and the app sees activity on the controller device it schedules a Status update on the Parent device.

**Reply by 64Spaces** [reply]
Thank you @THX723 and @ctgreybeard. I think you might be right, in that I'm not following the proper interpretation of "controllers". My goal was to update a Keypad Button (E) if a linked Wall Switch was turned off in the Home App, so they would stay in sync. I'm not able to get Controllers to work this way. But targetKeypad seems to provide some functionality for this.

**Reply by THX723** [reply]
@64Spaces For that you should be using Target Keypad LED parameters...

Target Keypad LED: By Insteon's design, keypad (scene button) LED follows the state of a linked scene only; it does not act according to the device state itself. eg. Turn on scene 1 then the corresponding keypad LED lights up, but turning on Light 1 directly will not light up the keypad LED. This enhancement allows you to specify which keypad LED(s) to set according to the device state; effetively turning keypad buttons into true device status indicators that many had wished for. For the following example, when Light 1 "XXYYZZ" is turned on (or at any dim level), button "A" of the 6-button keypad "AABBCC" is lit up, as do button "D" of the 8-button keypad "BBCCDD".

```
"name" : "Light 1",
"deviceID" : "XXYYZZ",
"targetKeypadID" : [ "AABBCC", "BBCCDD" ],
"targetKeypadBtn" : [ "A", "D" ],
"targetKeypadSixBtn" : [ true, false ]
```
